lib/flux-core/src/iter/traits/iterator.rs

```diff
@@ -0,0 +1,33 @@
+use flux_attrs::*;
+
+#[extern_spec(core::iter)]
+#[assoc(fn valid_item(self: Self, item: Self::Item) -> bool { true })]
+#[assoc(fn size(self: Self) -> int)]
+#[assoc(fn done(self: Self) -> bool)]
+#[assoc(fn step(self: Self, other: Self) -> bool { true })]
+trait Iterator {
+    #[flux::sig(fn(self: &strg Self[@curr_s]) -> Option<Self::Item>[!<Self as Iterator>::done(curr_s)] ensures self: Self{next_s: <Self as Iterator>::step(curr_s, next_s)})]
```

Comment:
I found this hard to read in a single line. I would format it like so. 

```suggestion
    #[flux::sig(
        fn(self: &strg Self[@curr_s]) -> Option<Self::Item>[!<Self as Iterator>::done(curr_s)] 
        ensures self: Self{next_s: <Self as Iterator>::step(curr_s, next_s)}
    )]
```